e2e/claim_test.go

```diff
@@ -0,0 +1,256 @@
+		Contract: minterAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      []byte(`{"mint":{}}`),
+		Funds:    sdk.NewCoins(sdk.NewInt64Coin("ustars", 100_000_000)),
+	})
+	require.NoError(t, err)
+
+	// Buyer should have 100STARS less
+	require.Equal(t,
+		sdk.NewInt64Coin("ustars", 1_900_000_000).String(),
+		app.BankKeeper.GetBalance(ctx, accs[1].Address, "ustars").String(),
+	)
+
+	claimRecords := []claimtypes.ClaimRecord{
+		{
+			Address:                addr1.String(),
+			InitialClaimableAmount: sdk.NewCoins(sdk.NewInt64Coin(claimtypes.DefaultClaimDenom, 1_000_000_000)),
+			ActionCompleted:        []bool{false, false, false, false, false},
+		},
+	}
+	err = app.ClaimKeeper.SetClaimRecords(ctx, claimRecords)
+	require.NoError(t, err)
+
+	instantiateRes, err = msgServer.InstantiateContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgInstantiateContract{
+		Sender: creator.Address.String(),
+		Admin:  creator.Address.String(),
+		CodeID: 2,
+		Label:  "Claim",
+		Msg:    []byte("{}"),
+	})
+	require.NoError(t, err)
+	require.NotNil(t, instantiateRes)
+	require.NotEmpty(t, instantiateRes.Address)
+	claimAddress := instantiateRes.Address
+	require.NotEmpty(t, claimAddress)
+
+	// claim airdrop
+	claimMsgTemplate := `
+		{
+		"claim_mint_nft" : {
+				"minter_address":"%s"
+			}
+		}
+	`
+	claimMsgRaw := []byte(fmt.Sprintf(claimMsgTemplate, minterAddress))
+	_, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgExecuteContract{
+		Contract: claimAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      claimMsgRaw,
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "address is not allowed to claim")
+
+	// allow contract
+
+	app.ClaimKeeper.SetParams(ctx, claimtypes.Params{
+		AirdropEnabled:     true,
+		AirdropStartTime:   startDateTime,
+		DurationUntilDecay: claimtypes.DefaultDurationUntilDecay,
+		DurationOfDecay:    claimtypes.DefaultDurationOfDecay,
+		ClaimDenom:         claimtypes.DefaultClaimDenom,
+		AllowedClaimers: []claimtypes.ClaimAuthorization{
+			{
+				ContractAddress: claimAddress,
+				Action:          claimtypes.ActionBidNFT,
+			},
+		},
+	})
+
+	claimMsgRaw = []byte(fmt.Sprintf(claimMsgTemplate, minterAddress))
+	_, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), &wasmtypes.MsgExecuteContract{
+		Contract: claimAddress,
+		Sender:   accs[1].Address.String(),
+		Msg:      claimMsgRaw,
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "address is not allowed to claim")
+	claim, err := app.ClaimKeeper.GetClaimRecord(ctx, addr1)
+	require.NoError(t, err)
+	require.False(t, claim.ActionCompleted[claimtypes.ActionMintNFT])
+
+	app.ClaimKeeper.SetParams(ctx, claimtypes.Params{
+		AirdropEnabled:     true,
+		AirdropStartTime:   startDateTime,
+		DurationUntilDecay: claimtypes.DefaultDurationUntilDecay,
+		DurationOfDecay:    claimtypes.DefaultDurationOfDecay,
+		ClaimDenom:         claimtypes.DefaultClaimDenom,
+		AllowedClaimers: []claimtypes.ClaimAuthorization{
```

Comment:
does this limits the contracts that minters can claim against?

Reply:
this limits an address that can submit the claim message to the chain. In this case after the contract is uploaded another gov proposal needs to be submitted with the contract address and only that instantiation can claim the airdrop.